### DISCUSSION TITLE
Multiple Compare keys in Blacklist Rule and Txt file format

### DISCUSSION THREAD

**Discussion by BeeruxBash** [jertel/elastalert2#1095] [Q&A] [answered]
Hey everyone,
I am trying to combine the two blacklist rules into one by trying to include multiple compare keys within a single, type: blacklist rule along with their blacklist parameters. But due to scare information available online even in the official documentation related to it I am not sure is it even possible to have multiple compare key arguments. 
As I tried several possibilities but either it returned 'oneof' error or it just executed the first compare key block and ignored the second one when I just placed them one after another.

Following are both the individual rules:

###  Blacklist Rule to detect an agent ip from a text file
```
name: IOC detection rule
type: blacklist
index: wazuh-alerts-*
description: This is an Elastalert triggered alert!!!
num_events: 01
timeframe:
    minutes: 01

compare_key: agent.ip

blacklist:
      - "!file /home/ubuntu/ips.txt"
alert:
- slack 
```

### Blacklist Rule to match the hash value
```
name: IOC detection rule 2
type: blacklist
index: wazuh-alerts-*
description: This is an Elastalert triggered alert!!!
num_events: 01
timeframe:
    minutes: 01

compare_key: data.win.eventdata.hashes

blacklist:
      - "!file /home/ubuntu/hashes.txt"
alert:
- slack
```
### Contents of the Text files:
```
ubuntu@ubuntu-virtual-machine:~$ cat ips.txt 
192.168.100.46
192.168.100.128
ubuntu@ubuntu-virtual-machine:~$ cat hashes.txt 
MD5=3388221800DAA1A0B3AFE08A0A6491A0,SHA256=32C940A939E234737C1F7BEC99A442EE8A05F38144B6AC30D1CD9820D366F086,IMPHASH=CDFCF001FC68E44FFF11F499922453F5
```
So, basically I am sending the alerts to Slack using its webhook URL.

### Combined Rule Attempt:
```
name: IOC detection rule 3
type: blacklist
index: wazuh-alerts-*
description: This is an Elastalert triggered alert!!!
num_events: 01
timeframe:
    minutes: 01

compare_key:
      - agent.ip

blacklist:
      - "!file /home/ubuntu/ips.txt"

compare_key:
        - data.win.eventdata.hashes

blacklist:
      - "!file /home/ubuntu/hashes.txt"

#filter:
#    - query:
#        query_string:
#          query: "data.win.eventdata.hashes:!file /home/ubuntu/hashes.txt OR agent.ip:!file /home/ubuntu/ips.txt"
#blacklist:
#      - "data.win.eventdata.hashes: !file /home/ubuntu/hashes.txt"
#      - "agent.ip: !file /home/ubuntu/ips.txt"
#blacklist:
#      - "!file /home/ubuntu/hashes.txt"
#      - "!file /home/ubuntu/ips.txt"
alert:
- slack
```

I tried searching ChatGPT to find a solution. It yielded the following result:

### ChatGPT's response to the same query
```
rules:
  - name: Example blacklist rule with two compare keys
    type: blacklist
    index: your-index-name
    compare_key:
      - key1
      - key2
    filter:
      - query:
          query_string:
            query: "field1:value1 AND field2:value2"
    blacklist:
      - field1: value1
        field2: value2
```
Here, I don,t understand what does the field 1 & 2 specify in the context of my rules.

Also, I am only able to match the 1st IP Address in the text file against the alerts and the not the 2nd one, I tried putting them in a list, place them in double quotations marks but still it wont match the 2nd entry. Is there any specific format for txt file too for using them like this? 
Similarly, I am unable to match a single hash value out of three MD5, SHA256 & IMPHASH present in the txt file. Basically, this is how these three values appear in the alert and if I just want to match only one of the for e.g. MD5 hash out of the three by specifiying its value in the txt file, I am unable to do it as it doesn't show an hits for it.
![image](https://user-images.githubusercontent.com/106651547/215559317-612243f3-a184-432e-a67c-58b9a46733c1.png)
Here are the values as they appear in the alert.
Can I use Regex Patterns to match these values, like is it supported?

Thanks In Advance !

**Comment by jertel** [comment, marked as answer]
Unfortunately, no, ElastAlert 2 does not support combining blacklist rules with conflicting compare_keys. Nor does it support regex comparisons. You could add an ingest rule to split the hashes field into hashes.md5, hashes.sha256, etc fields. If your source data and your blacklist files consistently supplies all three hashes then you really only need to compare one of them in the blacklist rule.

**Reply by BeeruxBash** [reply]
Hey jertel, just wanted to confirm is regex comparison supported in Blacklist & Whitelist rules? When comparing the compare_key value with the one stored in either Blacklist or Whitelist file? 
In my case, the compare_key is `data.win.eventdata.image` having values in this format `C:\\Program Files\\Rebex Buru SFTP Server\\burusftp.exe`. So, can I compare values coming from the compare key with the ones stored in the Blacklist or Whitelist by using regex to escape the file path and just match the executable name?

**Reply by jertel** [reply]
Hello @BeeruxBash, no the blacklist/whitelist logic is a simple `if x in list` statement, so it does not support regex. It would be a useful addition, but Elasticsearch would likely suffer from performance degradation if it needed to perform comparison's against 1000's of regex patterns for each document. If you research this further and find Elasticsearch has been optimized to support something like this, we can discuss it further.
